internal/commands/dockerdesktop.go

```diff
@@ -0,0 +1,153 @@
+package commands
+
+import (
+	"fmt"
+	"net/url"
+	"runtime"
+
+	"github.com/pkg/errors"
+
+	"github.com/docker/cli/cli/context/docker"
+	"github.com/docker/cli/cli/context/kubernetes"
+	"github.com/docker/cli/cli/context/store"
+	"github.com/docker/docker/client"
+	apiv1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/client-go/kubernetes/typed/core/v1"
+)
+
+type dockerDesktopHostProvider func() (string, bool)
+
+func defaultDockerDesktopHostProvider() (string, bool) {
+	switch runtime.GOOS {
+	case "windows", "darwin":
+	default:
+		// platforms other than windows or mac can't be Docker Desktop
+		return "", false
+	}
+	return client.DefaultDockerHost, true
+}
+
+type dockerDesktopLinuxKitIPProvider func() (string, error)
+
+type dockerDesktopDockerEndpointRewriter struct {
+	defaultHostProvider dockerDesktopHostProvider
+}
+
+func (r *dockerDesktopDockerEndpointRewriter) rewrite(ep *docker.EndpointMeta) {
+	defaultHost, isDockerDesktop := r.defaultHostProvider()
+	if !isDockerDesktop {
+		return
+	}
+	// on docker desktop, any context with host="" or host=<default host> should be rewritten as host="unix:///var/run/docker.sock" (docker socket path within the linuxkit VM)
+	if ep.Host == "" || ep.Host == defaultHost {
+		ep.Host = "unix:///var/run/docker.sock"
+	}
+}
+
+type dockerDesktopKubernetesEndpointRewriter struct {
+	defaultHostProvider dockerDesktopHostProvider
+	linuxKitIPProvider  dockerDesktopLinuxKitIPProvider
+}
+
+func (r *dockerDesktopKubernetesEndpointRewriter) rewrite(ep *kubernetes.EndpointMeta) {
+	// any error while rewriting makes as if no rewriting rule applies
+	if _, isDockerDesktop := r.defaultHostProvider(); !isDockerDesktop {
+		return
+	}
+	// if the kube endpoint host points to localhost or 127.0.0.1, we need to rewrite it to whatever is linuxkit VM IP is (with port 6443)
+	hostURL, err := url.Parse(ep.Host)
+	if err != nil {
+		return
+	}
+	hostName := hostURL.Hostname()
+	switch hostName {
+	case "localhost", "127.0.0.1":
```

Comment:
Dumb question but... What about ipv6 ? Shouldn't we also add it ? `::1` AFAIR

Reply:
Or maybe that's a linuxkit only thing, and only available with ipv4

Reply:
the kubeconfig file being generated by docker-desktop, I guess they will keep using localhost/127.0.0.1... but I can do it yes